lib/smart_proxy_remote_execution_ssh/api.rb

```diff
@@ -1,73 +1,116 @@
+      include Net::SSH::BufferedIo
+      extend Forwardable
+
+      # The list of methods taken from OpenSSL::SSL::SocketForwarder for the object to act like a socket
+      def_delegators(:@socket, :to_io, :addr, :peeraddr, :setsockopt, :getsockopt, :fcntl, :close, :closed?, :do_not_reverse_lookup=)
+
+      def initialize(socket)
+        @socket = socket
+        initialize_buffered_io
       end
-      def recv(n)
-        readpartial(n)
+
+      def resv
+        raise NotImplementedError
       end
-      def send(mesg, flags)
-        write(mesg)
+
+      def send
+        raise NotImplementedError
+      end
+
+      def self.applies_for?(socket)
+        raise NotImplementedError
+      end
+
+      def self.build(socket)
+        klass = [PumaBufferedSocket, OpenSSLBufferedSocket, StandardBufferedSocket].find do |potential_class|
+          potential_class.applies_for?(socket)
+        end
+        raise "No suitable implementation of buffered socket available for #{socket.inspect}" unless klass
+        klass.new(socket)
       end
     end
-  end
-end
 
-module OpenSSL
-  module SSL
-    class SSLSocket
+    class StandardBufferedSocket < BufferedSocket
+      def_delegators(:@socket, :send, :recv)
+
+      def self.applies_for?(socket)
+        socket.respond_to?(:send) && socket.respond_to?(:recv)
+      end
+    end
+
+    class OpenSSLBufferedSocket < BufferedSocket
+      def self.applies_for?(socket)
+        socket.is_a? ::OpenSSL::SSL::SSLSocket
+      end
+      def_delegators(:@socket, :read_nonblock, :write_nonblock, :close)
+
       def recv(n)
         res = ""
         begin
           # To drain a SSLSocket before we can go back to the event
           # loop, we need to repeatedly call read_nonblock; a single
           # call is not enough.
           while true
-            res += read_nonblock(n)
+            res += @socket.read_nonblock(n)
           end
         rescue IO::WaitReadable
           # Sometimes there is no payload after reading everything
           # from the underlying socket, but a empty string is treated
           # as EOF by Net::SSH. So we block a bit until we have
           # something to return.
           if res == ""
-            IO.select([io])
+            IO.select([@socket.to_io])
             retry
           else
             res
           end
         rescue IO::WaitWritable
           # A renegotiation is happening, let it proceed.
-          IO.select(nil, [io])
+          IO.select(nil, [@socket.to_io])
           retry
         end
       end
 
       def send(mesg, flags)
         begin
-          write_nonblock(mesg)
+          @socket.write_nonblock(mesg)
         rescue IO::WaitWritable
           0
         rescue IO::WaitReadable
-          IO.select([io])
+          IO.select([@socket.to_io])
           retry
         end
       end
     end
-  end
-end
 
-module Proxy::RemoteExecution
-  module Ssh
+    class PumaBufferedSocket < BufferedSocket
```

Comment:
As discussed, I would drop all pretends of supporting Puma in this PR.

Reply:
Updated